### PR TITLE
MDNS registry fix for users on VPNs

### PR DESCRIPTION
**Pull request by domwong**
or other software (firewalls etc) that may block MDNS responses. 

In some cases, if a user was using a VPN, the MDNS server would try to respond to the query using the public IP address of the machine i.e. as assigned by their VPN. In most cases the VPN server would block the response which meant that the client never received the packets and thus MDNS registry would never work. This means running `micro server` would fail.

The proposed fix is to detect if the IP address that we are responding to is the local machine and, if so, attempt to also send our responses to `0.0.0.0` thus avoiding the VPN. This is done on a best efforts basis with no error checking. We detect the IP at start up using a simple ping to Google's DNS server (perhaps we should pick a different address since this may not be available in all locations - e.g. China).

### CODE REVIEW

**Comment by sadwxqezc** [comment]
Google DNS is available in China

File: util/mdns/server.go [hunk 1]
```diff
@@ -430,13 +432,23 @@ func (s *Server) sendResponse(resp *dns.Msg, from net.Addr) error {
...
+	_, err = conn.WriteToUDP(buf, addr)
+	// If the address we're responding to is this machine then we can also attempt sending on 0.0.0.0
+	// This covers the case where this machine is using a VPN and certain ports are blocked so the response never gets there
+	// Sending two responses is OK
+	if addr.IP.Equal(s.outboundIP) {
+		// ignore any errors, this is best efforts
+		conn.WriteToUDP(buf, &net.UDPAddr{IP: backupTarget, Port: addr.Port})
 	}
+	return err
+
```
Comment (sadwxqezc) [inline util/mdns/server.go:466]: I think it's better to make sending on `0.0.0.0` an option

File: util/mdns/server.go [hunk 2]
```diff
@@ -474,3 +486,17 @@ func setCustomPort(port int) {
...
+	if err != nil {
+		// no net connectivity maybe so fallback
+		return nil
+	}
+	defer conn.Close()
+
+	localAddr := conn.LocalAddr().(*net.UDPAddr)
+
+	return localAddr.IP
+}
```
Comment (sadwxqezc) [inline util/mdns/server.go:517]: Better to make this an interface

**Comment by domwong** [comment]
> Google DNS is available in China

Thanks for confirming. 

Have taken onboard your other suggestions. Thanks for the input!

**Review by crufter — APPROVED** [review]

**Pull request merged**